### PR TITLE
Complete program execution on exceptions in async_mode=True

### PR DESCRIPTION
**Pull request by jprafael**
Fixes and adds unit test for #206. Uses the proposed implementation with extra stuff to clean-up tasks after completion.

### CODE REVIEW

File: guidance/_program.py [hunk 1]
```diff
@@ -306,9 +324,11 @@ def _stream_run(self, loop, new_program):
         yield new_program
 
         # cancel all tasks and close the loop
-        for task in asyncio.all_tasks(loop=loop):
+        for task in self._tasks:
             task.cancel()
         loop.run_until_complete(asyncio.sleep(0)) # give the loop a chance to cancel the tasks
+
+        # TODO: do we really want to close the loop? what if it is used by others?
```
Comment (jprafael) [inline guidance/_program.py:331]: @slundberg is it possible that the `loop` is shared with other usages and that this `loop.close()` should not be called (and be the caller's resposability to do that?)

**Review by yufei-realm — APPROVED** [review]

**Comment by jprafael** [comment]
@microsoft-github-policy-service agree company="Realm"

**Comment by slundberg** [comment]
Thanks! Will dig into this soon.

File: guidance/_program.py (hunk 1 above)
Comment (slundberg) [reply guidance/_program.py:331]: Great point.

**Comment by slundberg** [comment]
Thank you again for this PR @jprafael ! It looks good to me, and I am merging now.

**Pull request merged**